README.md

```diff
@@ -20,135 +14,123 @@ protocol ProtocolName {
-Valid arguments exist for both the auto-flattening and the monadic versions to be the default.
-Protocols eliminate this issue in two ways:
-
-1. Symbols are unique and unambiguous. There is no fear of naming collisions,
-   and it is clear what function you are using.
-1. Protocols may be applied to existing classes, so there is nothing
-   preventing consumers with different goals from using their own methods.
 
-```js
-protocol Functor {
-  map;
-}
-
-class Identity {
-  constructor(val) { this.val = val; }
-  unwrap() { return this.val; }
+  // Coherence-guaranteed implementation for existing classes
+  implFor Object {
+    tag = 'Object';
+  };
 }
-
-Promise.prototype[Functor.map] = function (f) {
-  return this.then(function(x) {
-    if (x instanceof Identity) {
-      x = x.unwrap();
-    }
-    let result = f.call(this, x);
-    if (result instanceof Promise) {
-      result = new Identity(result);
-    }
-    return result;
-  });
-};
-
-Protocol.implement(Promise, Functor);
 ```
 
-Finally, one of the biggest benefits of protocols is that they eliminate the
-fear of mutating built-in prototypes. One of the beautiful aspects of
-ECMAScript is its ability to extend its built-in prototypes. But with the
-limited string namespace, this is untenable in large codebases and impossible
-when integrating with third parties. Because protocols are based on symbols,
-this is no longer an anti-pattern.
+Protocols isolate implementations meant for different purposes by keeping them
+in a separate object with certain static guarantees.
+
+One of the biggest benefits of protocols is that they eliminate the
+fear of mutating built-in prototypes in potentially conflicting ways. One of the
+beautiful aspects of ECMAScript is its ability to extend its built-in
+prototypes. But with the limited string namespace, this is untenable in large
+codebases and impossible when integrating with third parties. Because protocols
+are based on symbols and support coherence, this is no longer an anti-pattern.
 
 ```js
 class Ordering {
-  static LT = new Ordering;
-  static EQ = new Ordering;
-  static GT = new Ordering;
+  constructor (str) { this.symbol = Symbol(str) }
 }
 
 protocol Ordered {
+  static LT = new Ordering('LT');
+  static EQ = new Ordering('EQ');
+  static GT = new Ordering('GT');
+  static fromNum (num) {
+      if (num < 0) return Ordered.LT;
+      if (num === 0) return Ordered.EQ;
+      if (num > 0) return Ordered.GT;
+    }
+  }
   compare;
 
   lessThan(other) {
-    return this[Ordered.compare](other) === Ordering.LT;
+    return this[Ordered].compare(other) === Ordered.LT;
+  }
+
+  implFor String {
+    compare(other) {
+      return Ordered.fromNum(String.localeCompare(other));
+    }
+  }
+
+  implFor Number {
+    compare(other) {...}
   }
 }
 
-String.prototype[Ordered.compare] = function() { /* elided */ };
-Protocol.implement(String, Ordered);
+(2)[Ordered].compare(1) // Ordered.GT
```

Comment:
Is this meant to be `[Ordered.compare]()`? Or is it the string proprety “compare” on an object under the Ordered key?

Reply:
The latter.

Reply:
What is that object? What’s it’s prototype, is it mutable, etc?

Reply:
:handwave: whatever it needs to be, in the end. Let's treat it as a question mark for now, because it will probably go away shortly anyway.